File: src/L01/Validation.hs
module L01.Validation where

type Err = String

--  class Validation<A> {
--    Validation(String error) {} // Error
--    Validation(A value) {} // Value
--  }
data Validation a = Error Err | Value a
  deriving (Eq, Show)

isError :: Validation a -> Bool
isError (Error _) = True
isError (Value _) = False

isValue :: Validation a -> Bool
isValue = not . isError

mapValidation :: (a -> b) -> Validation a -> Validation b
mapValidation _ (Error s) = Error s
mapValidation f (Value a) = Value (f a)

bindValidation :: Validation a -> (a -> Validation b) -> Validation b
bindValidation (Error s) _ = Error s
bindValidation (Value a) f = f a

valueOr :: Validation a -> a -> a
valueOr (Error _) a = a
valueOr (Value a) _ = a

errorOr :: Validation a -> Err -> Err
errorOr (Error e) _ = e
errorOr (Value _) a = a

valueValidation :: a -> Validation a
valueValidation = Value

File: src/L01/Validation.scala
package L01

import Validation._

sealed trait Validation[A] {
  def isError: Boolean =
    this match {
      case Error(_) => true
      case Value(_) => false
    }

  def isValue: Boolean =
    !isError

  def map[B](f: A => B): Validation[B] =
    this match {
      case Error(s) => Error(s)
      case Value(a) => Value(f(a))
    }

  def flatMap[B](f: A => Validation[B]): Validation[B] =
    this match {
      case Error(s) => Error(s)
      case Value(a) => f(a)
    }

  def valueOr(d: => A): A =
    this match {
      case Error(_) => d
      case Value(a) => a
    }

  def errorOr(a: => Err): Err =
    this match {
      case Error(e) => e
      case Value(_) => a
    }

}
case class Error[A](e: Err) extends Validation[A]
case class Value[A](a: A) extends Validation[A]

object Validation {
  type Err =
    String
}
